### PR TITLE
Add experimental ai/core features

### PR DESCRIPTION
**Pull request by lgrammel**
## Summary
- add `experimental_generateText`
- add `experimental_streamText`
- add `experimental_generateObject`
- add `experimental_streamObject`
- add language model specification
- add OpenAI chat and completion providers
- add Mistral chat provider

### CODE REVIEW

File: .changeset/lemon-beans-bathe.md [hunk 1]
```diff
@@ -2,4 +2,4 @@
 'ai': patch
 ---
 
-feat(ai/rsc): support string diff and patch in streamable value
+feat(ai/rsc): support string diff and patch in streamable value
```
Comment (lgrammel) [inline .changeset/lemon-beans-bathe.md:5]: Weird result of resetting the changesets from the canary branch.

File: packages/core/core/generate-text/stream-text.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  get fullStream(): AsyncIterableStream<TextStreamPart<TOOLS>> {
+    return createAsyncIterableStream(this.originalStream, {
+      transform(chunk, controller) {
+        if (chunk.type === 'text-delta') {
+          // do not stream empty text deltas:
+          if (chunk.textDelta.length > 0) {
+            controller.enqueue(chunk);
+          }
```
Comment (shuding) [inline packages/core/core/generate-text/stream-text.ts:135]: But it still looks like that the initial `textDelta` is always empty from the tests?

File: packages/core/core/util/is-deep-equal-data.ts
[the remark this answers is not in the document]
Comment (lgrammel) [reply packages/core/core/util/is-deep-equal-data.ts:1]: There is. This was re-implemented on purpose to reduce the dependencies we import (and it's a special case).

[the remark this answers is not in the document]
Comment (lgrammel) [reply packages/core/core/util/is-deep-equal-data.ts:1]: (I expect close to zero maintenance on this particular function).

File: packages/core/ai-model-specification/util/load-api-key.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+  if (typeof process === 'undefined') {
+    throw new LoadAPIKeyError({
+      message: `${description} API key is missing. Pass it using the '${apiKeyParameterName}' parameter. Environment variables is not supported in this environment.`,
+    });
+  }
+
+  apiKey = process.env[environmentVariableName];
+
+  if (apiKey == null) {
```
[the remark this answers is not in the document]
Comment (lgrammel) [reply packages/core/ai-model-specification/util/load-api-key.ts:32]: Done (same)

File: packages/core/core/generate-text/stream-text.ts (hunk 2 above)
Comment (lgrammel) [reply packages/core/core/generate-text/stream-text.ts:135]: It depends on the provider. Providers might have empty chunks, but we don't forward them.

File: packages/core/core/prompt/call-settings.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  /**
+   * The seed (integer) to use for random sampling. If set and supported
+   * by the model, calls will generate deterministic results.
+   */
+  seed?: number;
+
+  /**
+   * Maximum number of retries. Set to 0 to disable retries.
+   *
+   * Default: 2.
```
[the remark this answers is not in the document]
Comment (lgrammel) [reply packages/core/core/prompt/call-settings.ts]: done

File: packages/core/core/util/retry-with-exponential-backoff.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        message: `Failed after ${tryNumber} attemps. Last error: ${errorMessage}`,
+        reason: 'maxRetriesExceeded',
+        errors: newErrors,
+      });
+    }
+
+    if (error instanceof Error) {
+      if (error.name === 'AbortError') {
+        throw error;
+      }
```
[the remark this answers is not in the document]
Comment (lgrammel) [reply packages/core/core/util/retry-with-exponential-backoff.ts]: done

File: examples/ai-core/data/comic-cat.png
Comment (MaxLeiter) [inline examples/ai-core/data/comic-cat.png:1]: This is 1.2 MB. Can we switch to something a bit smaller?

Comment (lgrammel) [reply examples/ai-core/data/comic-cat.png:1]: Lol. I'll make it smaller.

**Review by shuding — APPROVED** [review]

File: packages/core/mistral/convert-to-mistral-chat-messages.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                }
+                case 'image': {
+                  throw new UnsupportedFunctionalityError({
+                    provider,
+                    functionality: 'image-part',
+                  });
+                }
+              }
+            })
+            .join(''),
```
[the remark this answers is not in the document]
Comment (shuding) [reply packages/core/mistral/convert-to-mistral-chat-messages.ts:40]: Gotcha!

File: packages/core/core/generate-text/generate-text.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+  // execute tools:
+  const toolResults =
+    tools == null ? [] : await executeTools({ toolCalls, tools });
+
+  return new GenerateTextResult({
+    // Always return a string so that the caller doesn't have to check for undefined.
+    // If they need to check if the model did not return any text,
+    // they can check the length of the string:
+    text: modelResponse.text ?? '',
```
Comment (MaxLeiter) [inline packages/core/core/generate-text/generate-text.ts:77]: If we _always_ want to return text, `||` might be safer

```suggestion
    text: modelResponse.text || '',
```

File: packages/core/core/generate-text/run-tools-transformation.ts [hunk 8]
```diff
@@ -0,0 +1,186 @@
+import { nanoid } from 'nanoid';
```
Comment (MaxLeiter) [inline packages/core/core/generate-text/run-tools-transformation.ts:1]: Do we still have our custom nanoid config in `shared/utils`? Should this use that?

**Review by MaxLeiter — APPROVED** [review]

File: packages/core/core/generate-text/generate-text.ts (hunk 7 above)
Comment (lgrammel) [reply packages/core/core/generate-text/generate-text.ts:77]: If it's not text, I'd rather throw an error. `||` might hide an error condition

**Comment by joblee2** [comment]
please add doc for these new feature(like `experimental_streamText`)

**Comment by lgrammel** [comment]
@joblee2 https://sdk.vercel.ai/docs/reference/ai-sdk-core

**Pull request merged**